File: design-pattern-java/app/src/main/java/design/pattern/java/patterns/behavioral/memento/TextEditor.java
package design.pattern.java.patterns.behavioral.memento;

/**
 * Memento Pattern Package Example
 */
public class TextEditor {
  private TextEditorOriginator textEditor;
  private java.util.List<TextEditorSnapshot> history;

  public TextEditor() {
    this.textEditor = new TextEditorOriginator();
    this.history = new java.util.ArrayList<>();
  }

  public String printText() {
    return this.textEditor.getState();
  }

  public TextEditor write(String value) {
    this.history.add(this.textEditor.save());
    this.textEditor.write(value);
    return this;
  }

  public TextEditor undo() {
    if (this.history.size() <= 0)
      return this;
    this.textEditor.restore(this.history.get(this.history.size() - 1));
    this.history = this.history.subList(0, this.history.size() - 1);
    return this;
  }
}

File: design-pattern-jvm/app/src/main/kotlin/design/pattern/kotlin/patterns/behavioral/memento/TextEditor.kt
package design.pattern.kotlin.patterns.behavioral.memento

/**
 * Memento Pattern Package Example
 */
class TextEditor {
    private val textEditor: TextEditorOriginator = TextEditorOriginator()
    private var history: MutableList<TextEditorSnapshot>
    fun printText(): String {
        return textEditor.state
    }

    fun write(value: String?): TextEditor {
        history.add(textEditor.save())
        textEditor.write(value!!)
        return this
    }

    fun undo(): TextEditor {
        if (history.size <= 0) return this
        textEditor.restore(history[history.size - 1])
        history = history.subList(0, history.size - 1)
        return this
    }

    init {
        history = mutableListOf()
    }
}